### PR TITLE
Add governor metadata bridge

### LINKED ISSUE #7: Create L2 GovernorMetadata Contract
#### Description

- Setup a contract that is able to read L1 state to get a valid start block for an L1 proposal, validate and L1 proposal and accept submitted votes from the `ERC20Votes token`.

### PR DESCRIPTION
**Pull request by alexkeating**
#### Description

- We add contracts to send governor contract metadata such as proposal start block and end block from L1 to L2
- We added deploy scripts and moved existing deploy scripts to the scripts directory

Closes #7

### CODE REVIEW

**Comment by alexkeating** [comment]
Current dependencies on/for this PR:
* main
  * **PR #11** <a href="https://app.graphite.dev/github/pr/ScopeLift/l2-flexible-voting/11" target="_blank">Graphite</a> 
    * **PR #8** <a href="https://app.graphite.dev/github/pr/ScopeLift/l2-flexible-voting/8" target="_blank">Graphite</a>  👈

This comment was auto-generated by [Graphite](https://app.graphite.dev/github/pr/ScopeLift/l2-flexible-voting/8?utm_source=stack-comment).

File: src/L1GovernorMetadataBridge.sol [hunk 1]
```diff
@@ -0,0 +1,55 @@
+// SPDX-License-Identifier: MIT
+pragma solidity ^0.8.16;
+
+import {IGovernor} from "openzeppelin/governance/Governor.sol";
+import {IWormhole} from "wormhole/interfaces/IWormhole.sol";
+
+contract L1GovernorMetadataBridge {
```
Comment (apbendi) [inline src/L1GovernorMetadataBridge.sol:7]: Nit: might be nice to have a general natspec comment quickly describing the purpose of the contract.

File: src/L1GovernorMetadataBridge.sol [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+  /// @notice The governor where proposals are fetched and bridged.
+  IGovernor public governor;
```
Comment (apbendi) [inline src/L1GovernorMetadataBridge.sol:9]: nit: can be immutable, right?

File: src/L1GovernorMetadataBridge.sol [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+
+  /// @notice The Wormhole core contract used to relay messages.
+  IWormhole public coreBridge;
```
Comment (apbendi) [inline src/L1GovernorMetadataBridge.sol:12]: nit: immutability

File: src/L1GovernorMetadataBridge.sol [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+
+  /// @notice The proposal id is an invalid proposal id.
+  error InvalidProposalId();
+
+  /// @param _governor The address of the L1 governor contract.
+  /// @param _core The address of the L1 core wormhole contract.
+  constructor(address _governor, address _core) {
+    governor = IGovernor(_governor);
+    coreBridge = IWormhole(_core);
+    nonce = 0;
```
Comment (apbendi) [inline src/L1GovernorMetadataBridge.sol:32]: nit: not really needed—initializes to 0 by default

File: src/WormholeReceive.sol [hunk 5]
```diff
@@ -0,0 +1,52 @@
+// SPDX-License-Identifier: MIT
+pragma solidity ^0.8.16;
+
+import {IWormhole} from "wormhole/interfaces/IWormhole.sol";
+
+abstract contract WormholeReceive {
```
Comment (apbendi) [inline src/WormholeReceive.sol:6]: Nice I like pulling out these utility messages. Should we also leverage this in the token bridging contracts, then?

**Review by apbendi — COMMENTED** [review]
A few comments after an initial quick pass through @alexkeating. Thanks!

File: src/L1GovernorMetadataBridge.sol (hunk 1 above)
Comment (alexkeating) [reply src/L1GovernorMetadataBridge.sol:9]: https://github.com/ScopeLift/l2-flexible-voting/pull/8/commits/187ca6f32075fc097c14cd1a81072cf5b245203a

File: src/L1GovernorMetadataBridge.sol (hunk 2 above)
Comment (alexkeating) [reply src/L1GovernorMetadataBridge.sol]: https://github.com/ScopeLift/l2-flexible-voting/pull/8/commits/5211bb94a9b648a59dbedd1b6cb22f116421c456

File: src/L1GovernorMetadataBridge.sol (hunk 3 above)
Comment (alexkeating) [reply src/L1GovernorMetadataBridge.sol]: https://github.com/ScopeLift/l2-flexible-voting/pull/8/commits/5f0bd1f05aae38f147e9e2c9883960cd50b33de2

File: src/L1GovernorMetadataBridge.sol (hunk 4 above)
Comment (alexkeating) [reply src/L1GovernorMetadataBridge.sol]: https://github.com/ScopeLift/l2-flexible-voting/pull/8/commits/d4a00f358908af70cf6016252ee2b4dab019d2f2

File: src/WormholeReceive.sol (hunk 5 above)
Comment (alexkeating) [reply src/WormholeReceive.sol]: Good suggestion added https://github.com/ScopeLift/l2-flexible-voting/pull/8/commits/b7f3ac5d29509fd00ae44607171b50d5f8df8b1f

File: script/RegisterL1Bridge.s.sol [hunk 6]
```diff
@@ -0,0 +1,28 @@
+// SPDX-License-Identifier: MIT
+pragma solidity ^0.8.16;
+
+import {Script, stdJson} from "forge-std/Script.sol";
+import {IERC20Receive} from "src/interfaces/IERC20Mint.sol";
+import {Constants} from "test/Constants.sol";
+
+contract RegisterL1Bridge is Script, Constants {
```
Comment (apbendi) [inline script/RegisterL1Bridge.s.sol:8]: Do you think for this and other scripts, we can add a few sentence comment at the top of each documenting a bit what's going on? Something along the lines of "This script should be run on the L2 by the owner of the token in order to register the blah blah blah as required by wormhole for blah blah blah" or whatever? I'm having a little trouble orienting myself as-to the purpose of each script, and those comments could go a long way toward making it more clear.

File: src/WormholeReceive.sol [hunk 7]
```diff
@@ -0,0 +1,53 @@
+// SPDX-License-Identifier: MIT
+pragma solidity ^0.8.16;
+
+import {IWormhole} from "wormhole/interfaces/IWormhole.sol";
+import {Ownable} from "openzeppelin/access/Ownable.sol";
+
+abstract contract WormholeReceive is Ownable {
```
[the remark this answers is not in the document]
Comment (alexkeating) [reply src/WormholeReceive.sol:7]: Agreed

File: script/RegisterL1Bridge.s.sol (hunk 6 above)
Comment (alexkeating) [reply script/RegisterL1Bridge.s.sol:8]: Yes, this would definitely be helpful

Comment (alexkeating) [reply script/RegisterL1Bridge.s.sol:9]: https://github.com/ScopeLift/l2-flexible-voting/pull/8/commits/fc68660afa4e4a251a9a47c3cf880ee88c44d58f

File: src/WormholeReceive.sol (hunk 7 above)
[the remark this answers is not in the document]
Comment (alexkeating) [reply src/WormholeReceive.sol]: https://github.com/ScopeLift/l2-flexible-voting/pull/8/commits/0257e707109c380dc7b150f10cefa55ffa43bafd

File: src/L1GovernorMetadataBridge.sol [hunk 8]
```diff
@@ -0,0 +1,56 @@
...
+  /// @param _governor The address of the L1 governor contract.
+  /// @param _core The address of the L1 core wormhole contract.
+  constructor(address _governor, address _core) {
+    GOVERNOR = IGovernor(_governor);
+    CORE_BRIDGE = IWormhole(_core);
+  }
+
+  /// @param l2GovernorMetadata The address of the L2 governor metadata contract.
+  function initialize(address l2GovernorMetadata) public {
+    if (!INITIALIZED) {
```
[the remark this answers is not in the document]
Comment (alexkeating) [reply src/L1GovernorMetadataBridge.sol:41]: https://github.com/ScopeLift/l2-flexible-voting/pull/8/commits/0257e707109c380dc7b150f10cefa55ffa43bafd

[the remark this answers is not in the document]
Comment (apbendi) [reply src/L1GovernorMetadataBridge.sol:41]: Was this nesting possibly a rebasing error? Doesn't make sense as-is. Should get rid of the !INITIALIZED check altogether.

File: test/L2GovernorMetadata.t.sol [hunk 9]
```diff
@@ -0,0 +1,30 @@
...
+    );
+  }
+}
+
+contract ReceiveEndcodedMsg is L2GovernorMetadataTest {
+  function testFork_CorrectlyStoreMetadata() public {
+    // Encoded message was generated from transaction
+    // 0x873dab653e1a8a746a8e4f51eb0e1de46f878255312e18506301e3cc5cb29954
+    // on Avalanche Fuji
+    l2GovernorMetadata.receiveEncodedMsg(
```
Comment (apbendi) [inline test/L2GovernorMetadata.t.sol:26]: Should I assume we are just setting the harness up here, and that more robust unit testing can/will be added later?

File: script/RegisterL1GovernorMetadataBridge.s.sol [hunk 10]
```diff
@@ -0,0 +1,30 @@
+// SPDX-License-Identifier: MIT
+pragma solidity ^0.8.16;
+
+import {Script, stdJson} from "forge-std/Script.sol";
+
+import {L2GovernorMetadata} from "src/L2GovernorMetadata.sol";
+import {Constants} from "test/Constants.sol";
+
+/// @dev This script will register the chain and address of the L1 metadata contract on the L2
```
Comment (apbendi) [inline script/RegisterL1GovernorMetadataBridge.s.sol:9]: Am I correct in saying this must be run by the `owner` address? Might just be good to note this in the comments.

**Review by apbendi — COMMENTED** [review]
A few additional comments @alexkeating, looking pretty good!

File: src/L1GovernorMetadataBridge.sol (hunk 8 above)
[the remark this answers is not in the document]
Comment (alexkeating) [reply src/L1GovernorMetadataBridge.sol:41]: Yes, that was probably what happened

[the remark this answers is not in the document]
Comment (alexkeating) [reply src/L1GovernorMetadataBridge.sol]: https://github.com/ScopeLift/l2-flexible-voting/pull/8/commits/2fecce979800cfa2d5276827fdb5e1bf8f8ccd09

File: test/L2GovernorMetadata.t.sol (hunk 9 above)
Comment (alexkeating) [reply test/L2GovernorMetadata.t.sol:26]: Yes, in depth testing will be in a separate pr

File: script/RegisterL1GovernorMetadataBridge.s.sol (hunk 10 above)
Comment (alexkeating) [reply script/RegisterL1GovernorMetadataBridge.s.sol:9]: Yes, that is correct and I will add that

Comment (alexkeating) [reply script/RegisterL1GovernorMetadataBridge.s.sol:9]: https://github.com/ScopeLift/l2-flexible-voting/pull/8/commits/e6d6ab6e53a69f70e9bb63ccba2ef916d186f9bf

**Pull request merged**